FirebaseFunctionsSwift/Sources/Codable/Callable+Codable.swift

```diff
@@ -0,0 +1,199 @@
+    -> Callable<Request, Response> {
+    return Callable(callable: httpsCallable(name), encoder: encoder, decoder: decoder)
+  }
+
+  /// Creates a reference to the Callable HTTPS trigger with the given name. The types of the `Encodable`
+  /// request and tyhe `Decodable` response will be inferred by the compiler.
+  ///
+  /// At the call site, use the following syntax:
+  ///
+  ///     let greeter: Callable<DataTestRequest, DataTestResponse> = functions.httpsCallable("greeter")
+  ///     try greeter(data) { result in
+  ///       print(result.greeting)
+  ///     }
+  ///
+  /// - Parameters:
+  ///   - name: The name of the Callable HTTPS trigger
+  func httpsCallable<Request, Response>(_ name: String,
+                                        encoder: StructureEncoder = StructureEncoder(),
+                                        decoder: StructureDecoder = StructureDecoder())
+    -> Callable<Request, Response> where Request: Encodable, Response: Decodable {
+    return Callable(callable: httpsCallable(name), encoder: encoder, decoder: decoder)
+  }
+}
+
+/**
+ * A `Callable` is reference to a particular Callable HTTPS trigger in Cloud Functions.
+ */
+public struct Callable<Request: Encodable, Response: Decodable> {
+  /**
+   * The timeout to use when calling the function. Defaults to 60 seconds.
+   */
+  public var timeoutInterval: TimeInterval {
+    get {
+      callable.timeoutInterval
+    }
+    set {
+      callable.timeoutInterval = newValue
+    }
+  }
+
+  enum CallableError: Error {
+    case internalError
+  }
+
+  private let callable: HTTPSCallable
+  private let encoder: StructureEncoder
+  private let decoder: StructureDecoder
+
+  init(callable: HTTPSCallable, encoder: StructureEncoder, decoder: StructureDecoder) {
+    self.callable = callable
+    self.encoder = encoder
+    self.decoder = decoder
+  }
+
+  /**
+   * Executes this Callable HTTPS trigger asynchronously.
+   *
+   * The data passed into the trigger must be of the generic `Request` type:
+   *
+   * The request to the Cloud Functions backend made by this method automatically includes a
+   * Firebase Instance ID token to identify the app instance. If a user is logged in with Firebase
+   * Auth, an auth ID token for the user is also automatically included.
+   *
+   * Firebase Instance ID sends data to the Firebase backend periodically to collect information
+   * regarding the app instance. To stop this, see `[FIRInstanceID deleteIDWithHandler:]`. It
+   * resumes with a new Instance ID the next time you call this method.
+   *
+   * - Parameter data: Parameters to pass to the trigger.
+   * - Parameter completion: The block to call when the HTTPS request has completed.
+   *
+   * - Throws: An error if any value throws an error during encoding.
+   */
+  public func call(_ data: Request,
+                   completion: @escaping (Result<Response, Error>)
+                     -> Void) throws {
```

Comment:
In the review thread of #8854 we discussed if it could make sense to 'squash' the two types of error handling into one:

https://github.com/firebase/firebase-ios-sdk/pull/8854#issuecomment-974697150

and @schmidt-sebastian answered that he thought it would be a good solution in this case.

Should I make a PR for that, or can you change the method to be non-throwing and rather have a `do {} catch {} around the entire function body that calls the completion with the error in the `catch` clause?

Reply:
Done in bdacb29c6.